feature(Report to Compare Invoiced Items to received Stock)

lomamech opened the pull request:
- Implement the report to display what the patient has billed and what
  have received and what distribution without invoice Attibution

closes #4380

lomamech committed:
feature(Report to compaare invoiced items to received stock)
- Implement the report to display what the patient has billed and what
have received and what distribution without invoice Attibution

Fixes #4380

lomamech committed:
Sanitaze : Code

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -1 +1,30 @@
+) AS m_p ON sm.uuid = m_p.uuid
+SET sm.period_id = m_p.period_id;
+
+-- @lomamech 2020-04-29
```
Everything in this file before this is in the previous migration script.  Can you delete everything above this and leave the new changes you made?

jniles replied:
This still needs to be fixed.

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/report.handlebars:
```diff
@@ -0,0 +1,148 @@
+            </tr>
+          </thead>
+          <tbody>
+            {{#each noInvoiceAttributionAggregat}}
```
If there isn't any extra stock not attributed to an invoice, can we hide this?

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/index.js:
```diff
@@ -0,0 +1,203 @@
+    JOIN inventory AS inv ON inv.uuid = itm.inventory_uuid
+    JOIN document_map AS map ON map.uuid = iv.uuid
+    WHERE p.uuid = ? AND (DATE(iv.date) >= DATE(?) AND DATE(iv.date) <= DATE(?)) 
+    AND inv.consumable = 1
```
:+1:

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/index.js:
```diff
@@ -0,0 +1,203 @@
+// default report parameters
+const DEFAULT_PARAMS = {
+  csvKey : 'invoiced_received_stock',
+  filename : 'REPORT.PATIENT_STANDING.DESCRIPTION',
```
This filename should reflect what the report is.  The `filename` property is used to name the file when exporting to PDF.

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/index.js:
```diff
@@ -0,0 +1,203 @@
+ *
+ * @description
+ * This report makes it possible to attach the items billed to a patient on
+ * the articulated items consumed or distributed to the Patient.
```
Good description :+1:

jniles reviewed (commented):
@lomamech, you put this together really quickly!  Awesome.

This looks really good.  A few comments to make the module easier to use:
  1. Could you order the inventory items in each invoice alphabetically?  This would make it easier to compare when someone doesn't give Quinine 300mg, but gives Quinine 500mg.  
  2.  If there is no data for an inventory item or an invoice item, could you leave the cell blank instead of putting 0?  That way, the user can see easily what is missing.  I've noted that in red below.
  3. If an invoice has no consumable items and there is no stock linked to the invoice, we should hide it.  There is no point in showing it.
  4. If there are no extra stock items, we should hide the last table that shows extra stock items.  It's in red below.
  5.  Could you put a subtotal at the bottom of each $$ column in the report?

![image](https://user-images.githubusercontent.com/896472/80946586-481bd700-8de6-11ea-8004-3087d239f07f.png)

lomamech committed:
refactor(Report to compare invoiced items to received stock)
- Set inventory items in invoice Alphabetically
- set cell blank instead of put 0
- Don't display invoice has no consumable items
- Set subtotal at the bottom of invoice

lomamech committed:
Resolve conflict with branch Master

lomamech committed:
fix(Report to compare invoiced items to received stock)
- Correction of error on the table
- Display of the difference in costs (invoicing on distribution) instead of
the difference in the cost of invoicing the item on compared to the cost
of distribution
- Correction in the joint concerning the unit cost of distribution

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/index.js:
```diff
@@ -0,0 +1,181 @@
+const DEFAULT_PARAMS = {
+  csvKey : 'invoiced_received_stock',
+  filename : 'REPORT.COMPARE_INVOICED_RECEIVED.TITLE',
+  orientation : 'landscape',
```
I think this report makes more sense in a `portrait` than landscape.  Here is an example of my tests.

![image](https://user-images.githubusercontent.com/896472/81284797-80295100-9056-11ea-9af5-5202b675f7a1.png)

jniles replied:
Or maybe just removing the padding would address this, actually.

lomamech replied:
> Or maybe just removing the padding would address this, actually.

I also just noticed that when you download a report, the shades of gray disappear, with you noticing this problem 
![image](https://user-images.githubusercontent.com/8035588/81390977-4454c100-9114-11ea-818f-c8bc536e2639.png)

jniles replied:
Yeah, that has to do with #4382.  It's why I was hoping for a solution that didn't involve `!important;` everywhere.  Don't worry about that for now.

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/index.js:
```diff
@@ -0,0 +1,181 @@
+    JOIN inventory AS inv ON inv.uuid = l.inventory_uuid
+    JOIN invoice AS iv ON iv.uuid = sm.invoice_uuid
+    JOIN document_map AS map ON map.uuid = iv.uuid
+    WHERE sm.invoice_uuid IS NOT NULL AND
```
Do you need this `sm.invoice_uuid IS NOT NULL`?  I think the `JOIN` on inventory should already take care of that, no?

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/report.handlebars:
```diff
@@ -0,0 +1,176 @@
+<html>
+  {{> head title="REPORT.COMPARE_INVOICED_RECEIVED.TITLE"}}
+  <body>
+  <div class="container">
```
Can you make this `container-fluid` to not have as much padding?

jniles commented on server/controllers/finance/reports/invoicedReceivedStock/report.handlebars:
```diff
@@ -0,0 +1,176 @@
+            <td class="text-right"><strong> {{ debcred total_movement ../../metadata.enterprise.currency_id }}  </strong></td>
+          </tr>
+          <tr>
+            <td colspan="5"></td>
```
Why do you have these extra `<td>`s?  They add a kind of weird three lines below the report.  Could you just put a `padding` or `margin` on something to add space if you need it?  Or even a `<br>`?

![image](https://user-images.githubusercontent.com/896472/81285907-21fd6d80-9058-11ea-8fad-be5f44c9aa11.png)

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -8,4 +8,34 @@ INSERT INTO unit VALUES
+-- @jeremielodi 23-04-2020
+-- stock consumption graph report
+
+INSERT INTO unit VALUES
```
The only thing that should be in this file is the `INSERT INTO flux` from @mbayopanda and your additions:
```sql
INSERT INTO unit VALUES
  (270, 'compare invoiced to received','TREE.COMPARE_INVOICED_RECEIVED','Compare invoiced items to received stock', 144,'/modules/reports/invoicedReceivedStock','/reports/invoicedReceivedStock');

INSERT INTO `report` (`report_key`, `title_key`) VALUES
  ('invoicedReceivedStock', 'REPORT.COMPARE_INVOICED_RECEIVED.TITLE');
```

Please make these changes.

jniles reviewed (commented):
@lomamech this is really close!  I've tested it on live data and it works well, the best I can tell.

I've made a few comments.  The `migration/next.sql` is really important to address before a merge.  It will break upgrades if we don't address it.

lomamech committed:
improve(Report to compare invoiced items to received stock)
- Clean migration/next.sql
- Clean sql, delete not necessary condition
- Remove not necessary extra line in table of report

jniles reviewed (approved):
This works really well.  Thanks @lomamech !

bors r+
